[RFC] Add a proposal for synchronous garbage collection

caesarxuchao opened the pull request:
Tracking issue: #29891

@lavalamp @kubernetes/sig-api-machinery @kubernetes/api-review-team

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32628)

<!-- Reviewable:end -->

Commit:
add a proposal for synchronous garbage collection

smarterclayton commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,59 @@
+* When deleting dependents, it should use the same `DeleteOptions.SynchronousGC` as the owner’s finalizers suggest.
+
+**Note**
+We cannot have circular dependencies anymore. Otherwise SynchronousGC will enter a deadlock.
```
How will we handle this?

lavalamp replied:
Why would this be the case. It should not be the case.

smarterclayton commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,59 @@
+We cannot have circular dependencies anymore. Otherwise SynchronousGC will enter a deadlock.
+
+## Unhandled cases
+* If the GC observes the owning object with the GCFinalizer before it observes the creation of all the dependents, GC will remove the finalizer from the owning object before all dependents are gone. Hence, “Synchronous GC” is best-effort, though we guarantee that the dependents will be deleted eventually.
```
Add a section for implications of changes to clients like namespace controller, node controller, kubelet, etc.

smarterclayton commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,59 @@
+
+**Modifications to processItem()**
+
+* Treat owner whose DeletionTimestamp is non-nil as “not exist”.
```
What's the motivation for this?  More detail here please

Commit:
addressing comments

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,73 @@
+
+**Modifications to processItem()**
+
+`processItem()` consumes the `dirtyQueue`, requests the API server to delete an item if all of its owners do not exist. To support synchronous GC, it has to:
```
Added more detail as requested.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,73 @@
+* If the GC observes the owning object with the GCFinalizer before it observes the creation of all the dependents, GC will remove the finalizer from the owning object before all dependents are gone. Hence, “Synchronous GC” is best-effort, though we guarantee that the dependents will be deleted eventually. We face a similar case when handling OrphanFinalizer, see [GC known issues](https://github.c ...
+
+
+## Implications to existing clients
```
Added the section as requested. @smarterclayton what exactly do you want to see in this section? I don't think synchronous GC will affect kubelet or node controller.

smarterclayton replied:
Just want to make sure that we think about how they impact it.  I.e. the namespace controller ended up needing a tweak as we just found.  Components that expect to delete and have it go away will be surprised if it is deleted and doesn't go away.  I.e. the kubelet sync loop runs twice and keeps trying to delete the pod (or fails because it expects the pod to be gone).

For node controller we eventually do a force deletion.  How will force deletion affect the finalizer?  If synchronous GC blocks due to a circular reference, what happens to the node eviction loop (does it hang / go pathological)?  Just those sorts of questions

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,73 @@
+
+**Handling circular dependencies**
+
+SynchronousGC will enter a deadlock in the presence of circular dependencies. To break the deadlock, we need to timeout a GCFinalizer. To implement the timeout, GC adds an object that has a GCFinalizer into a [delaying queue](../../pkg/util/workqueue/delaying_queue.go) when it's observed, and removes the GCFinalizer from it when the time is up. The timeout value should be proportional to the numb ...
```
I think we can handle circular dependencies by enforcing a timeout on synchronous GC.

smarterclayton replied:
Does that make it possible to DDOS the server in an unusual / hard to debug way?

lavalamp replied:
Chao and I talked about some strategies for handling this. I don't think a timeout is a good solution.

Commit:
more

Commit:
addressing comments

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+**kubectl**: synchronous GC can replace the **kubectl delete** reapers. Currently `kubectl delete` blocks until all dependents and the owner are deleted. To maintain this behavior, after switched to using synchronous GC, *kubectl delete* needs to poll on the removal of the owner object.
+
+## Security implications
```
@liggitt @deads2k might be interested in this.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+**Handling circular dependencies**
+
+SynchronousGC will enter a deadlock in the presence of circular dependencies. The garbage collector can break the circle by lazily detecting circular dependencies: when `processItem()` processes an object, if it finds the object and all of its owners have the `GCFinalizer`, it searches the internal owner-dependency relationship graph (`uidToNode`) to check if the object and any of its owner are i ...
```
@lavalamp how about handling dependency circle this way? This should break a circle timely.

lavalamp replied:
I think it should track which object a user requested the sync deletion for, and which objects have the finalizer merely so that the GC can track their deletion (you could do this with two different finalizer strings, for example). The only thing that's important is that the thing the user sent gets deleted after everything else.

The only time there's a problematic circle is if users simultaneously delete all of the items in the circle. In that case the GC probably needs to downgrade someone.

lavalamp replied:
I added an idea at the top that expands on this concept of two finalizer classes.

lavalamp replied:
I talked with Brian about this, too-- since we don't expect users to do this on purpose, let's do the easiest possible thing and break the cycle as soon as we notice it (i.e., as soon as we add the sync finalizer to an object which depends on an object that already has the sync label, we remove that link). We can add guarantees later if needed.

caesarxuchao replied:
Thanks. I didn't mention the "two finalizers" idea in the proposal. I think we need only one finalizer for now, and we can switch to use two finalizers when we want to support circular dependencies.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+* If the GC observes the owning object with the `GCFinalizer` before it observes the creation of all the dependents, GC will remove the finalizer from the owning object before all dependents are gone. Hence, “Synchronous GC” is best-effort, though we guarantee that the dependents will be deleted eventually. We face a similar case when handling OrphanFinalizer, see [GC known issues](https://github ...
+
+
+## Implications to existing clients
```
Added analysis of several controllers that might be affected by synchronous GC.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+  // If SynchronousGarbageCollection is set, the object will not be deleted immediately. Instead, a GarbageCollectionInProgress finalizer will be placed on the object. The garbage collector will remove the finalizer from the object when all depdendents are deleted.
+  // SynchronousGarbageCollection and OrphanDependents are exclusive.
+  // SynchronousGarbageCollection default to false.
+  // SynchronousGarbageCollection is cascading, i.e., the object’s dependents will be deleted with the same SynchronousGarbageCollection.
```
I don't think this is necessary, actually. It's one possible implementation.

lavalamp replied:
And it may be very undesirable as it makes it harder to handle cycles.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+  …
+  // If SynchronousGarbageCollection is set, the object will not be deleted immediately. Instead, a GarbageCollectionInProgress finalizer will be placed on the object. The garbage collector will remove the finalizer from the object when all depdendents are deleted.
+  // SynchronousGarbageCollection and OrphanDependents are exclusive.
+  // SynchronousGarbageCollection default to false.
```
We should consider defaulting it in the way that causes the smallest change in old client behavior.

lavalamp replied:
(might be different for different objects)

caesarxuchao replied:
Defaulting to false is the least disruptive way. Old clients don't expect GC.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+# Overview
+
+Some users of the server-side garbage collection need to tell if the garbage collection is done ([example](https://github.com/kubernetes/kubernetes/issues/19701#issuecomment-236997077)). Synchronous Garbage Collection is a best-effort (see [unhandled cases](#unhandled-cases)) mechanism to enable such use cases: after the API server receives a deletion request of an owning object, the object keeps ...
```
An idea might be to offer two flavors of sync GC:
* Class one merely waits for all dependent objects to be in terminating state. There's no circular dependency problems with this class.
* Class two waits for all dependent objects to additionally have their names released for reuse. There's a potential circular dependency problem here; one resolution is to say that the garbage collector may downgrade a single item in the circle to class one. A better fix is to allow releasing the name as a thing that happens before the object is physically deleted from the storage layer, but we currently can't do that. If we could, this wouldn't have circular dependency problems, either.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+## Security implications
+
+A user who is authorized to update one object can affect the synchronous GC behavior of another object. Specifically, by setting an object as a pod's owner, and setting a very long grace termination period for the pod, a user can make the synchronous GC of the owner to take long time.
```
This is potentially a big deal. Do we have any strategies to mitigate this? @deads2k

deads2k replied:
> This is potentially a big deal. Do we have any strategies to mitigate this? @deads2k

I think this is a general problem with finalizing behavior which this proposal makes more obvious, but which always existed.  Reading through, this looks like a way to provide nicer client behavior, but the mechanism is roughly the same.  I don't know that you can do much better than allowing a timeout and a timeout behavior (fail or orphan), along with a status that links to offenders.

It looks very similar to the case where someone adds a finalizer ref on a namespace, but fails to start a controller that collects that finalizer.  The namespace sticks around forever.  @smarterclayton and I have talked about it, but we've never come up with a general solution to the stuck finalizer or stuck initializer problem.

lavalamp replied:
I talked a bit with Brian and came up with the idea of the GC impersonating the user at deletion time, and if the object is not visible to that user, then ignoring or breaking the link. It seems like the GC controller can do something in this area.

caesarxuchao replied:
The GC impersonating idea doesn't solve the scenario I described:
User A only has access to object Oa of kind Ka, but he sets Oa's ownerRef to point to Ob of kind Kb. User A also makes Oa's deletion take a very long time.
When user B, who has access to both Ka and Kb, initiates the synchronous deletion of Ob, he will find Ob's deletion takes a very long time, because of Oa.
The problem in this scenario is that user A gets the ability to affect the behavior of Ob.

caesarxuchao replied:
> along with a status that links to offenders

I think the garbage collector needs to expose the dependencies via the controller manager for debugging purposes. I'm not sure about the security implications. I think it's safe if the exposed information only contains the resource kind and the UID.

caesarxuchao replied:
> I don't know that you can do much better than allowing a timeout and a timeout behavior (fail or orphan), along with a status that links to offenders.

I'm not in favor of a timeout. https://github.com/kubernetes/kubernetes/pull/33799#discussion_r81331718 suggested let users define a timeout. It's hard for user to estimate how long GC will take, because it's hard for a user to figure out how many dependents are out there. On the other hand, if GC defines a default timeout, user will never know if all dependents are actually deleted.

I think we should disallow a user from setting ownerReference to point to a resource that he doesn't have update/delete permission. I think that's possible in our current system?

lavalamp replied:
I think there's a number of ways to reduce the severity of this problem. These are just some ideas.

1. apiserver could control the list of finalizers you're allowed to add to an object. You'd have to register a finalizer and only sysadmins would have permission to do that. This reduces the chance of an attack where you keep an object around by adding a bogus finalizer.
2. If we go with the finalizer registry idea, you could supply a max time along with the finalizer (probably bad idea for a number of reasons).
3. We could make adding a finalizer require a particular permission, although this would be a field level auth. We might need a dedicated subresource for this.
4. We could consider adding a restriction, like "cross-user links are {not permitted by apiserver | not honored by sync deletion by the GC}"
5. We could add a restriction like, you can't add a link to an object that can't see you. Although this is pretty weird because it means less permissions is better for the attackee.
6. Maybe better than that, just require mutual visibility.
7. Actually, even better-- make the sync GC mode do a permission check. If Ob->Oa, and Oa is sync deleted, but Oa cannot see Ob, then still go ahead and delete Ob, but don't consider it as something that needs to be waited for. This still lets people in some namespace get their thing turned down when a dependency is turned down, but they can't block the dependency's turndown process.
8. We could surface a metric, time spent in sync GC. sysadmins could get paged if it starts going up. (we should surface this metric no matter what, arguably.)

lavalamp replied:
9. We should come up with a way to express what's blocking the deletion, maybe in an annotation. After N seconds, we add an annotation like "deletion waiting for: eviluser/pods/sleepforever to be deleted". Then a user could either manually remove that link or force delete it to make the original operation proceed.

caesarxuchao replied:
1, 3 might be useful even out of the scope of synchronous GC.
4 is not possible today, we don't record which user owns an object.
5 and 7 do not solve https://github.com/kubernetes/kubernetes/pull/32628#discussion_r80347466.
6 not sure what does "visibility" mean. If it means having create/delete/update permissions, then it's similar to what I suggested in https://github.com/kubernetes/kubernetes/pull/32628#discussion_r81382123. But that might block some useful use cases ( e.g., a use wants his pod to share the fate of a database service, but he doesn't have delete permission of the database service, so he can't setup the reference).

8,9 look reasonable.

lavalamp reviewed (changes requested):
Good start, but I have some suggestions.

erictune commented:
Didn't get a chance to read this but discussed in hall with @bgrant0607 and @lavalamp .

I think the following can happen:

1. User creates object Oa of kind Ka and gets 200 / verifies existence of Oa.
2. User creates object Ob of kind Kb that has parent object Oa.
3. GC controller is watching Ka's and Kb's.
4. GC controller sees Ob before it sees Oa.  GC Controller concludes Ob needs to be deleted.

I don't know any way to fix this with our consistency model.

caesarxuchao commented:
> GC controller sees Ob before it sees Oa. GC Controller concludes Ob needs to be deleted.

@erictune when handling Ob, GC issues a GET to check if Oa exists.

liggitt commented:
The Get can return a NotFound error in a HA etcd cluster.

caesarxuchao commented:
@liggitt I don't understand HA etcd cluster well. In the case you described, is it because one etcd has processed the creation of the object, but the consensus of the etcd cluster is that the object does not exist?

lavalamp commented:
We should do consistent reads, so that shouldn't be true.

On Fri, Sep 23, 2016 at 2:05 PM, Chao Xu <notifications@github.com> wrote:

> @liggitt <https://github.com/liggitt> I don't understand HA etcd cluster
> well. In the case you described, is it because one etcd has processed the
> creation of the object, but the consensus of the etcd cluster is that the
> object does not exist?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32628#issuecomment-249302203>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglsu59U0PMI0ZP9iz6kqdZBUNXCpfks5qtD8kgaJpZM4J8Saa>
> .
>

smarterclayton commented:
The consistent read is not guaranteed to go to the same cluster - if you
have an extension API reference using a different etcd (or sharded etcd)
you are broken.

(unless I misread the scenario described, in which case I'll go back to
staring at the ocean more)

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
````diff
@@ -0,0 +1,126 @@
+}
+```
+
+We will introduce a new standard finalizer: const GCFinalizer string = “GarbageCollectionInProgress”
````
The string for the OrphanFinalizer is "orphan" ([here](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L2307.)), so probably "collectingGarbage". I hope we had named the  OrphanFinalizer "orphaning".

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+`processEvent()` manages GC’s internal owner-dependency relationship graph, `uidToNode`. It updates `uidToNode` according to the Add/Update/Delete events in the cluster. To support synchronous GC, it has to:
+
+* handle Add or Update events where `obj.Finalizers.Has(GCFinalizer) && obj.DeletionTimestamp != nil`. The object will be added into the `synchronousGC queue`. The object will be marked as “GC in progress” in `uidToNode`.
```
We can. When processing the queue, GC needs to check what kind of item it is, and apply different operations. I think that will make the code hard to reason, so I prefer using different queue. Are you worrying about performance? Or are you worrying about synchronization problem?

lavalamp replied:
You can write a switch statement and call different methods. I do indeed think it's easier to reason about the concurrency properties if there's only one queue.

(Note that you'll be writing this if statement either way, either deciding what queue to put something in or deciding what action to take. I don't agree that it's easier to reason about the code if you do this if statement everywhere you decide which queue to put something in-- I actually expect that will lead to code duplication.)

Commit:
addressing more comments

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+Delete() function needs to check the DeleteOptions.SynchronousGarbageCollection.
+
+* The option is ignored if DeleteOptions.OrphanDependents is true or nil.
```
Done

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+**Node controller** forcefully deletes pod if the pod is scheduled to a node that does not exist ([code](../../pkg/controller/node/nodecontroller.go#L474)). The pod will continue to exist if it has pending finalizers. The node controller will futilely retry the deletion. Also, the `node controller` forcefully deletes pods before deleting the node ([code](../../pkg/controller/node/nodecontroller.g ...
+
+**Podgc** deletes terminated pods if there are too many of them in the cluster. `Podgc` should remove any pending finalizers to make sure the pods are deleted.
```
Updated.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+**Podgc** deletes terminated pods if there are too many of them in the cluster. `Podgc` should remove any pending finalizers to make sure the pods are deleted.
+
+**Deployment controller** adopts existing `ReplicaSet` (RS) if its template matches. If a matching RS has a pending `GCFinalizer`, deployment shouldn't adopt it, because the RS controller will not scale up/down a RS that's being deleted. Hence, `deployment controller` needs to check if a RS is being deleted before adopting it. If the RS is being deleted, then the `deployment controller` should wa ...
```
Updated.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+**PetSet controller** will be blocked by a pod with pending finalizers, so Synchronous GC might slow down its progress.
+
+**kubectl**: synchronous GC can replace the **kubectl delete** reapers. Currently `kubectl delete` blocks until all dependents and the owner are deleted. To maintain this behavior, after switched to using synchronous GC, *kubectl delete* needs to poll on the removal of the owner object.
```
Done. Old kubectls shouldn't be affected.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,140 @@
+
+**kubectl**: synchronous GC can simplify the **kubectl delete** reapers. Let's take the `deployment reaper` as an example, since it's the most complicated one. Currently, the reaper finds all `RS` with matching labels, scales them down, polls until `RS.Status.Replica` reaches 0, deletes the `RS`es, and finally deletes the `deployment`. If using the synchronous GC, `kubectl delete deployment` is a ...
+
+Note that this **changes the behavior** of `kubectl delete`. The command will be blocked until all pods are deleted from the key-value store, instead of being blocked until pods are in the terminating state. This means `kubectl delete` blocks for longer time, but it has the benefit that the resources used by the pods are released when the `kubectl delete` returns.
```
@lavalamp @janetkuo @deads2k if `kubectl delete` uses synchronous GC, its semantics will be changed slightly. Thoughts?

janetkuo replied:
What if the users choose to turn GC off? They can't use `kubectl delete` anymore.

lavalamp replied:
Kubectl will need a way to know if the GC is on or not, or we'll have to say that 1.5+ kubectl is no longer compatible with <= 1.3 apiserver.

deads2k replied:
> @lavalamp @janetkuo @deads2k if kubectl delete uses synchronous GC, its semantics will be changed slightly. Thoughts?

I'd like a `--wait` flag, perhaps even on by default since I think that matches user expectation better, but I don't have a significant objection.

caesarxuchao replied:
Ok. If `--wait==false`, kubectl delete uses `DeleteDependentsInBackground` and returns immediately; if `--wait==true`, then kubectl uses `DeleteAfterDependentsAreDeleted`, and polls until the object is gone.

caesarxuchao commented:
Pushed another version. PTAL. Thanks.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,126 @@
+
+* treat an owner as "not exist" if `owner.DeletionTimestamp != nil && !owner.Finalizers.Has(OrphanFinalizer)`, otherwise Synchronous GC will not progress because the owner keeps existing in the key-value store.
+* when deleting dependents, it should use the same `DeleteOptions.SynchronousGC` as the owner’s finalizers suggest.
+* if an object has multiple owners, some owners still exit while other owners are in the synchronous GC stage, then according to the existing logic of GC, the object wouldn't be deleted. To unblock the synchronous GC of those owners, `processItem()` has to remove the ownerReferences pointing to them.
```
Done.

gmarek commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,140 @@
+
+**kubelet** deletes the pod from the key-value store after all its containers are terminated ([code](../../pkg/kubelet/status/status_manager.go#L441-L443)). It also assumes that if the API server does not return an error, the pod is removed from the key-value store. Breaking the assumption will not break `kubelet` though, because the `pod` must have already been in the terminated `phase`, `kubele ...
+
+**Node controller** forcefully deletes pod if the pod is scheduled to a node that does not exist ([code](../../pkg/controller/node/nodecontroller.go#L474)). The pod will continue to exist if it has pending finalizers. The node controller will futilely retry the deletion. Also, the `node controller` forcefully deletes pods before deleting the node ([code](../../pkg/controller/node/nodecontroller.g ...
```
Deleting Pods without Nodes will be moved to PodGC #32495

caesarxuchao replied:
Thanks. I quickly skimmed the PR it seems the same argument applies to the logic in the PodGC.

gmarek replied:
Yup - it was a minor nit.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,140 @@
+
+Delete() function needs to check the `DeleteOptions.SynchronousGarbageCollection`.
+
+* The request is rejected with 400 if both `DeleteOptions.SynchronousGarbageCollection` and `DeleteOptions.OrphanDependents` are true (or default to true, see [DefaultGarbageCollectionPolicy](https://github.com/kubernetes/kubernetes/blob/release-1.4/pkg/registry/generic/registry/store.go#L500)).
```
If either one is set, I think we should make the most generous assumption possible about the other one. e.g. if you specify sync=true and don't specify orphan at all, we should assume you mean orphan=false, even if the resource would ordinarily default to orphaning.

caesarxuchao replied:
Updated.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,140 @@
+
+**Replication controller manager**, **Job controller**, and **ReplicaSet controller** ignore pods in terminated phase, so pods with pending finalizers will not block these controllers.
+
+**PetSet controller** will be blocked by a pod with pending finalizers, so Synchronous GC might slow down its progress.
```
...but that may be desirable. @erictune

lavalamp reviewed (changes requested):
I would still greatly prefer a single queue. No major changes other than that.

caesarxuchao commented:
Github randomly hides the reply area for some comments..

> Kubectl will need a way to know if the GC is on or not, or we'll have to say that 1.5+ kubectl is no longer compatible with <= 1.3 apiserver.

This [doc](https://github.com/kubernetes/kubernetes/blob/master/docs/design/versioning.md#supported-releases-and-component-skew) says we support version skew within 1 minor version, so we don't need to worry about that.

@janetkuo made a valid point that user can manually disable GC, and `kubectl delete` will then stop working. Can we say that's a user fault? To me, the case is analogous to that RC won't work if RC controller is disabled by user.

lavalamp commented:
We should remove the GC on/off flag in 1.5, I think it is a mandatory
system component.

On Tue, Sep 27, 2016 at 3:02 PM, Chao Xu <notifications@github.com> wrote:

> Github randomly hides the reply area for some comments..
>
> Kubectl will need a way to know if the GC is on or not, or we'll have to
> say that 1.5+ kubectl is no longer compatible with <= 1.3 apiserver.
>
> This doc
> <https://github.com/kubernetes/kubernetes/blob/master/docs/design/versioning.md#supported-releases-and-component-skew>
> says we support version skew within 1 minor version, so we don't need to
> worry about that.
>
> @janetkuo <https://github.com/janetkuo> made a valid point that user can
> manually disable GC, and kubectl delete will then stop working. Can we
> say that's a user fault? To me, the case is analogous to that RC won't work
> if RC controller is disabled by user.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32628#issuecomment-250012448>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngll6cidIvIHHGp013DGyOIvsWoXdxks5quZJogaJpZM4J8Saa>
> .
>

caesarxuchao commented:
> This doc says we support version skew within 1 minor version, so we don't need to worry about that.

I take it back. 1.4 server and garbage collector doesn't understand synchronous GC, so 1.5 kubectl needs to change its behavior based on that. 

Offline lavalamp and I discussed exposing a feature list at API server's `/version` endpoint. That's worth a standalone design doc. I'll write it before converting kubectl to use synchronous GC.

Commit:
API server makes generous assumption about DeleteOptions.OrphanDependents and
DeleteOptions.SynchronousGarbageCollection;

using only one queue;

address kubectl backward compatibility.

caesarxuchao reviewed (commented):
@lavalamp PTAL. Thanks.

caesarxuchao commented:
All my inline comments were lost by github..

caesarxuchao reviewed (commented):
@lavalamp PTAL. Thanks.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,141 @@
+To make the new kubectl compatible with the 1.4 and earlier masters, kubectl needs to switch to use the old reaper logic if it finds Synchronous GC is not supported by the master.
+
+Old kubectl is compatible with new master, because `DeleteOptions.SynchronousGarbageCollection` defaults to false.
+
```
@lavalamp updated. This is a little hand-waving. I'll figure out the details when converting kubectl to use synchronous GC.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,141 @@
+* when deleting dependents, it should use the same `DeleteOptions.SynchronousGC` as the owner’s finalizers suggest.
+* if an object has multiple owners, some owners still exist while other owners are in the synchronous GC stage, then according to the existing logic of GC, the object wouldn't be deleted. To unblock the synchronous GC of owners, `processItem()` has to remove the ownerReferences pointing to them.
+
+In addition, if an object popped from `dirtyQueue` is marked as "GC in progress", `processItem()` treats it specially:
```
@lavalamp updated to use one queue.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,141 @@
+Currently `processEvent()` manages GC’s internal owner-dependency relationship graph, `uidToNode`. It updates `uidToNode` according to the Add/Update/Delete events in the cluster. To support synchronous GC, it has to:
+
+* handle Add or Update events where `obj.Finalizers.Has(GCFinalizer) && obj.DeletionTimestamp != nil`. The object will be added into the `dirtyQueue`. The object will be marked as “GC in progress” in `uidToNode`.
+* Upon receiving the deletion event of an object, put its owner into the `dirtyQueue` if the owner node is marked as "GC in progress". This is to force the `processItem()` (described next) to re-check if all dependents of the owner is deleted.
```
@lavalamp updated to use one queue.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,141 @@
+Delete() function needs to check the `DeleteOptions.SynchronousGarbageCollection`.
+
+* The request is rejected with 400 if both `DeleteOptions.SynchronousGarbageCollection` and `DeleteOptions.OrphanDependents` are true.
+* If `DeleteOptions.SynchronousGarbageCollection` is explicitly set to true and `DeleteOptions.OrphanDependents` is nil, the API server will default `DeleteOptions.OrphanDependents` to false, regardless of the [default orphaning policy](https://github.com/kubernetes/kubernetes/blob/release-1.4/pkg/registry/generic/registry/store.go#L500) of the resource.
```
@lavalamp updated.

caesarxuchao reviewed (commented):
@lavalamp PTAL. Thanks.

deads2k commented:
> I'm not in favor of a timeout. #33799 (comment) suggested let users define a timeout. It's hard for user to estimate how long GC will take, because it's hard for a user to figure out how many dependents are out there. On the other hand, if GC defines a default timeout, user will never know if all dependents are actually deleted.

The user isn't trying to predict how long it will take, he's trying to prevent his script from getting stuck forever waiting for a delete.  If we return a status of "took too long", the caller/script has an option to come back with "delete an orphan".  Without that feature, what's our recommendation?

> I think we should disallow a user from setting ownerReference to point to a resource that he doesn't have update/delete permission. I think that's possible in our current system?

It is possible to inspect the create/update to determine if the user has those powers.

lavalamp commented:
Re: the synchronization problem. We can fix this by including in the owner ref the RV *of the dependent object*. LIST already accepts such a thing, GET could be modified to accept an RV and not return until that RV has been observed.

caesarxuchao commented:
> The user isn't trying to predict how long it will take, he's trying to prevent his script from getting stuck forever waiting for a delete. If we return a status of "took too long", the caller/script has an option to come back with "delete and orphan".

There is a misunderstanding here, probably caused by the word "Synchronous". The delete API call is not synchronous, it returns immediately. The owner object keeps existing. User can write a script to poll for the owner object, and knows all dependents are deleted when the GET of owner object receives a 404.

> Without that feature, what's our recommendation?

In the script they written, user can timeout the polling of the owner object, and manually remove the GCFinalizer from the owner, then the owner object will be deleted from the key-value store immediately.
(We probably need to add a "forceful deletion" delete option anyway, which will remove all the pending finalizers, set the grace-period to 0. This is a replacement of the current --grace-period=0 delete option in the finalizer era).

Commit:
add OwnerReference.BlockSynchronousGC

caesarxuchao commented:
@deads2k @lavalamp I pushed the commit "add OwnerReference.BlockSynchronousGC". It's a mechanism that prevents an evil user from blocking the synchronous GC. Could you take another look? Thanks.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -56,18 +57,33 @@ DeleteOptions {
+    ...
+    // Should the owner be deleted from the key-value store after this object is deleted during synchronous garbage collection.
+    // If the user creates the OwnerReference has delete permission of the owner, BlockSynchronousGC defaults to true; otherwise the field defaults to false, also 422 will be returned if the field is set to false.
+    BlockSynchronousGC *bool
```
`Synchronize` as a name for this. And I think this is an alternative design. You wouldn't do this AND the finalizer, it's one or the other. The idea with this is that the links are auth checked at admission time rather than GC time.

Actually I think you should add a `DeletionStrategy` field, with "background" (default) or "synchronous" as the possible values. It's bad to just drop bools in the API.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -56,18 +57,33 @@ DeleteOptions {
 * If `DeleteOptions.SynchronousGarbageCollection` is explicitly set to true and `DeleteOptions.OrphanDependents` is nil, the API server will default `DeleteOptions.OrphanDependents` to false, regardless of the [default orphaning policy](https://github.com/kubernetes/kubernetes/blob/release-1.4/pkg/registry/generic/registry/store.go#L500) of the resource.
 * If the option is set, the API server will update the object instead of deleting it, add the finalizer, and set the `ObjectMeta.DeletionTimestamp`.
 
+`validation.ValidateObjectMeta()` function needs to validate `OwnerReference.BlockSynchronousGC`. It needs to query the `Authorizer` to check if the user has delete permission of the owner object.
```
This is problematic-- shouldn't the permission check go the other way?

also auth checks should not be done in validation.

caesarxuchao commented:
An interesting idea: instead of introducing a new delete option, we introduce a new field, "Synchronous", to the `OwnerReference`. Authorization check is performed when an ownerReference is created, `Synchronous` must be `false` if the user doesn't have delete permission of the owner. This prevents the user from blocking the deletion of the owner.
 API server always puts a "CollectingGarbage" finalizer on the object when a deletion request is received, the garbage collector removes the finalizer when all dependents with `Synchronous==true` are deleted. Effectively, this moves the decision of whether to synchronously delete a group of objects from deletion time to creation (or update) time.
The disadvantage of this approach is that a deletion always requires two requests to the API server: one to start the deletion, another one to remove the "CollectingGarbage" finalizer. I don't see any easy optimization.
Another controversial thing is whether the user who issues the delete request should have the option to delete the object immediately, rather than having to respect the "Synchronous" ownerReferences set by other users.

Alternatively, we can keep the `SynchronousGC` delete option, if it's true, the object is deleted after all dependents with `Synchronous==true` ownerReference are deleted (note that a user who doesn't have delete permission of the owner object cannot create `Syncronous==true` ownerReference, so he cannot block the deletion of the owner); if it's false, than the object is deleted immediately.
The disadvantage is that the API is quite complex, having both the `SynchronousGC` delete option and the `Synchronous` field in `OwnerReference` is confusing.

@erictune @lavalamp @smarterclayton @deads2k what do you think?

lavalamp commented:
I thought of a way to get around needing two deletion requests, which is
for the creator to add the sync finalizer to the parent resource at
creation time. If we're already saying this is a mode the creator sets up,
then I think that's not too much of a hardship.

On Mon, Oct 3, 2016 at 8:36 PM, Chao Xu <notifications@github.com> wrote:

> An interesting idea: instead of introducing a new delete option, we
> introduce a new field, "Synchronous", to the OwnerReference.
> Authorization check is performed when an ownerReference is created,
> Synchronous must be false if the user doesn't have delete permission of
> the owner. This prevents the user from blocking the deletion of the owner.
> API server always puts a "CollectingGarbage" finalizer on the object when
> a deletion request is received, the garbage collector removes the finalizer
> when all dependents with Synchronous==true are deleted. Effectively, this
> moves the decision of whether to synchronously delete a group of objects
> from deletion time to creation (or update) time.
> The disadvantage of this approach is that a deletion always requires two
> requests to the API server: one to start the deletion, another one to
> remove the "CollectingGarbage" finalizer. I don't see any easy optimization.
> Another controversial thing is whether the user who issues the delete
> request should have the option to delete the object immediately, rather
> than having to respect the "Synchronous" ownerReferences set by other users.
>
> Alternatively, we can keep the SynchronousGC delete option, if it's true,
> the object is deleted after all dependents with Synchronous==true
> ownerReference are deleted (note that a user who doesn't have delete
> permission of the owner object cannot create Syncronous==true
> ownerReference, so he cannot block the deletion of the owner); if it's
> false, than the object is deleted immediately.
> The disadvantage is that the API is quite complex, having both the
> SynchronousGC delete option and the Synchronous field in OwnerReference
> is confusing.
>
> @erictune <https://github.com/erictune> @lavalamp
> <https://github.com/lavalamp> @smarterclayton
> <https://github.com/smarterclayton> @deads2k <https://github.com/deads2k>
> what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32628#issuecomment-251287825>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglsp9cJl6Qsm3dGp1NcbvbF5iMlV_ks5qwcm9gaJpZM4J8Saa>
> .
>

caesarxuchao commented:
If the owner reference is created manually, is the user required to manually add the finalizer for the owner? That implies the user needs update permission of the owner (which might not be wrong).

lavalamp commented:
Yes, I think that goes a long way towards mitigating the auth concerns...

On Tue, Oct 4, 2016 at 11:20 AM, Chao Xu <notifications@github.com> wrote:

> If the owner reference is created manually, is the user required to
> manually add the finalizer for the owner? That implies the user needs
> update permission of the owner (which might not be wrong).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32628#issuecomment-251470355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglqHF86PbhcX_vnXXLzLe7cZkUL9Nks5qwpkIgaJpZM4J8Saa>
> .
>

liggitt commented:
>> If the owner reference is created manually, is the user required to manually add the finalizer for the owner? That implies the user needs update permission of the owner (which might not be wrong).
>
> Yes, I think that goes a long way towards mitigating the auth concerns...

what's the difference between a "manually created" owner reference and a "controller created" owner reference?

caesarxuchao commented:
For example, when a replicaset controller creates pods, it sets the ownerReference of the pods to point back to the replicaset. That's what I meant by "controller created owner reference".

OTOH, `ObjectMeta.OwnerReferences` is just an API field, use can manually set reference, either via CREATE or PUT.

liggitt commented:
right, but those both look the same from an API perspective (or are you saying the controller would add the finalizer, and manual users would also be required to add the finalizer)

caesarxuchao commented:
> are you saying the controller would add the finalizer, and manual users would also be required to add the finalizer.

Yes.

caesarxuchao commented:
To sum up, the new idea is to let the user who creates the ownerReference to decide whether the deletion of the owner needs to wait for deletion of the dependent, via the `OwnerReference.Synchronous` field. The benefit is that the API server can check if the user has delete && update permission of the owner, if not, he cannot set `OwnerReference.Synchronous` to true. This prevents a user from blocking the deletion of the owner if he doesn't have delete && update of the owner.

A hardship for the user is that apart from setting the ownerReferences, he will also need to add the "GCFinalizer" to the owner. This is to avoid needing [two requests for a deletion](https://github.com/kubernetes/kubernetes/pull/32628#issuecomment-251461890). But given that most ownerReferences are set by various controllers, it's not too bad.

@erictune @deads2k @smarterclayton do you have a second opinion? Thanks.

Commit:
adding alternative design

caesarxuchao commented:
cc @fabioy who might have concerns from GKE perspective.

caesarxuchao commented:
cc @krousey

deads2k commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,205 @@
+
+SynchronousGC will enter a deadlock in the presence of circular dependencies. The garbage collector can break the circle by lazily breaking circular dependencies: when `processItem()` processes an object, if it finds the object and all of its owners have the `GCFinalizer`, it removes the `GCFinalizer` from the object.
+
+Note that the approach is not rigorous and thus having false positives. For example, if a user first sends a SynchronousGC delete request for an object, then sends the delete request for its owner, then `processItem()` will be fooled to believe there is a circle. We expect user not to do this. We can make the circle detection more rigorous if needed.
```
Even if we don't make it more rigorous, can we be sure that the implementation yields so that the controller doesn't get stuck?

caesarxuchao replied:
The approach only has false positives, meaning the garbage collector breaks the links even if there is no circle, so there won't be a deadlock.

deads2k commented on docs/proposals/synchronous-garbage-collection.md:
````diff
@@ -0,0 +1,205 @@
+}
+```
+
+Note that setting `BlockOwnerDeletion` alone is not enough, user also needs to add the "CollectingGarbage" finalizer to the owner. Considering most ownerReferences are set by controllers (e.g., replicaset controller), we think the burden is acceptable.
````
Wouldn't I expect many (most?) managing objects to have `"CollectingGarbage"` set by a controller, so I could just name it my parent and set this flag?  The controller does the work of setting the owning object finalizer for me and I just get to block it.

caesarxuchao replied:
Yeah, usually users don't need to manually set the "CollectingGarbage" finalizer.

deads2k commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,205 @@
+
+## Implications to existing clients
+
+The implications are mostly the same as Design I. One difference is that it causes behavior change of old version `kubectl delete`. Old version kubectl issues the delete request for the owner object after all dependent objects are terminating. An old version API server will delete the owner from the key-value store immediately, but a new version API server will keep the owner object around until  ...
```
This assumes that we only have a single client.  `kubectl` is certainly the most common, but I'm not sure we can reasonably claim its the only one for which we support backwards compatibility.

liggitt replied:
Agree, kubectl-centric code (like [this](https://github.com/kubernetes/kubernetes/blob/e72f26a3fff4f521d1da9f52b3ce82d4d0d5a401/pkg/registry/generic/registry/store.go#L275-L277)) in the API server should not exist. Working around compatibility problems that way makes us more likely to accept breaking clients that have integrated with the documented API.

caesarxuchao replied:
Ok. Thanks for pointing this out. I think I need to add back the DeleteOptions after all. I'll update the proposal.

caesarxuchao replied:
Pushed a new version. User sets `DeleteOptions.GarbageCollectionPolicy=GarbageCollectionSynchronous` to trigger synchronous GC. The owning object will be deleted after all dependents whose `OwnerReference.BlockOwnerDeletion=true` are deleted. This design is compatible with old version clients.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,205 @@
+
+### API Server
+
+When validating the ownerReference, API server needs to query the `Authorizer` to check if the user has "update" and "delete" permission of the owner object. It returns 422 if the user does not have the permissions.
```
@deads2k @liggitt do you have concerns on calling `Authorizer` in validation?

deads2k replied:
> @deads2k @liggitt do you have concerns on calling Authorizer in validation?

For "update", I would expect this to be handled during admission and I have no concern about doing that.

For "delete" its a little less clear since you don't have access to the original object until you get down into storage.  I think you'll probably want to wire up an admission check similar to what was done for update.

Commit:
Use DeleteOptions.GarbageCollectionPolicy and OwnerReference.BlockOwnerDeletion

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
````diff
@@ -0,0 +1,201 @@
+## DeleteOptions
+
+```go
+DeleteOptions {
````
Redesigned the DeleteOptions to make it easier to understand.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,201 @@
+| GarbageCollectionDefault      | OrphanDependents==nil    |
+| GarbageCollectionOrphan       | *OrphanDependents==true  |
+| GarbageCollectionAsynchronous | *OrphanDependents==false |
+| GarbageCollectionSynchronous  | N/A                      |
```
Do you prefer calling the two GC modes as Asynchronous/Synchronous or Background/Foreground?

lavalamp replied:
I vote for background/foreground.

deads2k commented:
> Regarding admission controllers, I think they are a bunch of plugins. I don't think we should write an admission controller whose sole responsibility is checking ownerReferences, but I don't see which existing admission controller can take this responsibility. Do you have suggestions?

I would use a separate admission plugin.  We have a proposal to ease admission, particularly the ordering aspects which are extremely difficult to reason about today (https://github.com/kubernetes/kubernetes/pull/34131/files#diff-2be0f77c0f5171454aa4db95bd66d342R110).  Given a structure like that, its pretty easy to add yourself to the chain in an appropriate location.

If you like, I can create one for owner references that requires permission to delete an object before allowing someone to set any ownerrefs to be used as a starting point.

caesarxuchao commented:
> If you like, I can create one for owner references that requires permission to delete an object before allowing someone to set any ownerrefs to be used as a starting point.

Thanks, that would be great, though we need to get both proposals in first.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
````diff
@@ -0,0 +1,201 @@
+)
+```
+
+`DeleteOptions.OrphanDependents *bool` will be removed. We decided not to add a `DeleteOptions.SynchronousGC *bool`, because together with `DeleteOptions.OrphanDependents` it results in 9 possible combinations and is thus confusing. This is a breaking change, so the new DeleteOptions will live in `api.k8s.io` group with version `v1`, as proposed in [#33900](https://github.com/kubernetes/kubernete ...
````
@smarterclayton @lavalamp can we make breaking changes to DeleteOptions and move it to `api.k8s.io/v1` in 1.5?

lavalamp replied:
Don't remove the old field, just add the new field and declare that only one of them may be set.

Mark the old field deprecated and say we'll remove it in 1.7.

It's easy to look at both fields and figure out which one the client set, and do the right thing.

caesarxuchao replied:
Done.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,201 @@
+  …
+  // Whether and how garbage collection will be performed.
+  // Defaults to GarbageCollectionDefault
+  GarbageCollectionPolicy GarbageCollectionPolicy
```
Make it a pointer--it's optional.

caesarxuchao replied:
Done.

smarterclayton commented:
I really don't like calling this Garbage Collection for end users.  It's
really not garbage, it's dependent object deletion.  I feel pretty strongly
we should be aligning what this *does*, not that it *is*

On Oct 11, 2016, at 3:01 PM, Daniel Smith <notifications@github.com> wrote:

*@lavalamp* commented on this pull request.
------------------------------

In docs/proposals/synchronous-garbage-collection.md
<https://github.com/kubernetes/kubernetes/pull/32628#pullrequestreview-3774163>
:

> +}
+```
+
+The initial draft of the proposal did not include this field and it
had a security loophole: a user who is only authorized to update one
resource can set ownerReference to block the synchronous GC of other
resources. Requiring users to explicitly set `BlockOwnerDeletion`
allows the master to properly authorize the request.
+
+## DeleteOptions
+
+```go
+DeleteOptions {
+  …
+  // Whether and how garbage collection will be performed.
+  // Defaults to GarbageCollectionDefault
+  GarbageCollectionPolicy GarbageCollectionPolicy
+}
+
+type GarabgeCollectionPolicy string

s/Garabge/Garbage/

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/32628#pullrequestreview-3774163>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p6frJo0FtY4zORmN4slpu8zsyGt3ks5qzAdHgaJpZM4J8Saa>
.

lavalamp commented:
How about

```
PropagationPolicy DeletionPropagationPolicy
```

as a name?

On Tue, Oct 11, 2016 at 3:12 PM, Clayton Coleman <notifications@github.com>
wrote:

> I really don't like calling this Garbage Collection for end users. It's
> really not garbage, it's dependent object deletion. I feel pretty strongly
> we should be aligning what this *does*, not that it *is*
>
> On Oct 11, 2016, at 3:01 PM, Daniel Smith <notifications@github.com>
> wrote:
>
> *@lavalamp* commented on this pull request.
> ------------------------------
>
> In docs/proposals/synchronous-garbage-collection.md
> <https://github.com/kubernetes/kubernetes/pull/32628#pullrequestreview-
> 3774163>
> :
>
> > +}
> +```
> +
> +The initial draft of the proposal did not include this field and it
> had a security loophole: a user who is only authorized to update one
> resource can set ownerReference to block the synchronous GC of other
> resources. Requiring users to explicitly set `BlockOwnerDeletion`
> allows the master to properly authorize the request.
> +
> +## DeleteOptions
> +
> +```go
> +DeleteOptions {
> + …
> + // Whether and how garbage collection will be performed.
> + // Defaults to GarbageCollectionDefault
> + GarbageCollectionPolicy GarbageCollectionPolicy
> +}
> +
> +type GarabgeCollectionPolicy string
>
> s/Garabge/Garbage/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32628#pullrequestreview-
> 3774163>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_
> p6frJo0FtY4zORmN4slpu8zsyGt3ks5qzAdHgaJpZM4J8Saa>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32628#issuecomment-253062474>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglvXEKdZaNuQ9KVet-6ChhCxdJ6z6ks5qzAnEgaJpZM4J8Saa>
> .
>

caesarxuchao commented:
How about this?

```go
DeleteOptions {
  …
  DependentsDeletionPolicy GarbageCollectionPolicy
}

type GarbageCollectionPolicy string

const (
    DefaultDependentsDeletionPolicy GarbageCollectionPolicy = "Default"
    OrphanDependents GarbageCollectionPolicy = "OrphanDependents"
    DeleteDependentsInBackground GarbageCollectionPolicy = "DeleteDependentsInBackground"
    DeleteAfterDependentsAreDeleted GarbageCollectionPolicy = "DeleteAfterDependentsAreDeleted"
)
```

Commit:
addressing comment; rename user facing "Garbage Collection"

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,201 @@
+  …
+  // Whether and how garbage collection will be performed.
+  // Defaults to DefaultPropagationPolicy
+  DeletionPropagationPolicy *DeletionPropagationPolicy
```
@smarterclayton how about these new names?

caesarxuchao replied:
Done.

lavalamp replied:
The field name can probably just be `PropagationPolicy`, it's redundant to say "delete" since it's in the DeleteOptions already. (Type name should still include "Delete")

caesarxuchao replied:
> The field name can probably just be PropagationPolicy

Done.

smarterclayton commented:
Probably "deletePropagationPolicy" is most correct.

On Tue, Oct 11, 2016 at 9:26 PM, Chao Xu <notifications@github.com> wrote:

> *@caesarxuchao* commented on this pull request.
> ------------------------------
>
> In docs/proposals/synchronous-garbage-collection.md
> <https://github.com/kubernetes/kubernetes/pull/32628>:
>
> > +     // Defaults to false.
> +     // To set this field, a user needs "update" and "delete" permission of the owner, otherwise 422 (Unprocessable Entity) will be returned.
> +     BlockOwnerDeletion *bool
> +}
> +```
> +
> +The initial draft of the proposal did not include this field and it had a security loophole: a user who is only authorized to update one resource can set ownerReference to block the synchronous GC of other resources. Requiring users to explicitly set `BlockOwnerDeletion` allows the master to properly authorize the request.
> +
> +## DeleteOptions
> +
> +```go
> +DeleteOptions {
> +  …
> +  // Whether and how garbage collection will be performed.
> +  // Defaults to DefaultPropagationPolicy
> +  DeletionPropagationPolicy *DeletionPropagationPolicy
>
> Do you mean "Delete" is present tense and "Deletion" is sort of perfect
> tense? I'll leave this to native speakers.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32628>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p6H_dt-Y6btx9Tf_VEb4OYo48hItks5qzGF2gaJpZM4J8Saa>
> .
>

caesarxuchao commented:
@krousey @lavalamp @deads2k are there more comments? This a 1.5 feature. Thanks.

deads2k commented:
> @krousey @lavalamp @deads2k are there more comments? This a 1.5 feature. Thanks.

I agree with @smarterclayton's nit on the name ("delete propagation" instead of "deletion propagation").

lgtm

Commit:
fixing nits; adding --wait to kubetl delete

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,203 @@
+
+**kubectl**: synchronous GC can simplify the **kubectl delete** reapers. Let's take the `deployment reaper` as an example, since it's the most complicated one. Currently, the reaper finds all `RS` with matching labels, scales them down, polls until `RS.Status.Replica` reaches 0, deletes the `RS`es, and finally deletes the `deployment`. If using synchronous GC, `kubectl delete deployment` is as ea ...
+
+Note that this **changes the behavior** of `kubectl delete`. The command will be blocked until all pods are deleted from the key-value store, instead of being blocked until pods are in the terminating state. This means `kubectl delete` blocks for longer time, but it has the benefit that the resources used by the pods are released when the `kubectl delete` returns. To allow kubectl user not waitin ...
```
@deads2k I added `--wait` to `kubectl delete`

lavalamp commented:
I'd argue that it should be "deletionOptions" and "deletionPropagationPolicy" but it's hard to change the former, so we might as well be consistent and say "deletePropagationPolicy".

Commit:
change the field name to PropagationPolicy

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,203 @@
+DeleteOptions {
+  …
+  // Whether and how garbage collection will be performed.
+  // Defaults to DefaultPropagationPolicy
```
// Either this field or OrphanDependents may be set, but not both.

caesarxuchao replied:
Done.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,203 @@
+
+const (
+    // Respects the existing garbage collection related finalizers on the object and the default garbage collection policy of the resource.
+    DefaultPropagationPolicy DeletePropagationPolicy = "Default"
```
DeletePropagationPolicyDefault
DeletePropagationPolicyOrphan
DeletePropagationPolicyBackground
DeletePropagationPolicyForeground

...or make the names consistent some other way. Maybe the word "policy" isn't required.

lavalamp replied:
In the comment, note explicitly that the default depends on the object type and the controller for the object. (If the controller pre-adds the sync finalizer label, then you'll get foreground behavior as the default unless you specifically request something else).

caesarxuchao replied:
Done.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,203 @@
+    // The object exists in the key-value store until the garbage collector deletes all the dependents whose ownerReference.blockOwnerDeletion=true from the key-value store.
+    // API sever will put the "DeletingDependents" finalizer on the object, and sets its deletionTimestamp.
+    // This policy is cascading, i.e., the dependents will be deleted with GarbageCollectionSynchronous.
+    DeleteAfterBlockingDependentsAreDeleted DeletePropagationPolicy = "DeleteAfterBlockingDependentsAreDeleted"
```
I think this name is not going to work. I prefer something short and not immediately obvious (like Foreground) so that users will read the longer, hopefully very clear, description in the documentation.

lavalamp commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,203 @@
+
+## API Server
+
+`Delete()` function checks `DeleteOptions.PropagationPolicy`. If the policy is `DeleteAfterBlockingDependentsAreDeleted`, the API server will update the object instead of deleting it, add the finalizer, and set the `ObjectMeta.DeletionTimestamp`.
```
What if the finalizer is already present?

lavalamp replied:
(i.e., if it's present and they ask for orphan or background gc, we need to remove it)

lavalamp reviewed (commented):
comments mostly on the api.

caesarxuchao commented on docs/proposals/synchronous-garbage-collection.md:
```diff
@@ -0,0 +1,204 @@
+type DeletePropagationPolicy string
+
+const (
+    // The default depends on the existing finalizers on the object and the type of the object.
```
> In the comment, note explicitly that the default depends on the object type and the controller for the object. (If the controller pre-adds the sync finalizer label, then you'll get foreground behavior as the default unless you specifically request something else).

@lavalamp Is the current comment clear? I think saying the default depends on the "existing finalizers" is more general than saying it depends on "the controller for the object".

caesarxuchao replied:
Also, do we need a `DeletePropagationDefault`? Can we use nil to mean default?

caesarxuchao replied:
Ok, I'll keep the default for now. We can move the comment to `DeleteOptions.PropagationPolicy` if we later decide `DeletePropagationDefault` is redundent.

lavalamp commented:
Comment looks good. Leaving a default might be good just to have a place to
put documentation, but I don't feel super strongly.

On Wed, Oct 12, 2016 at 4:23 PM, Chao Xu <notifications@github.com> wrote:

> *@caesarxuchao* commented on this pull request.
> ------------------------------
>
> In docs/proposals/synchronous-garbage-collection.md
> <https://github.com/kubernetes/kubernetes/pull/32628#pullrequestreview-3981287>
> :
>
> > +
> +## DeleteOptions
> +
> +```go
> +DeleteOptions {
> +  …
> +  // Whether and how garbage collection will be performed.
> +  // Defaults to DeletePropagationDefault
> +  // Either this field or OrphanDependents may be set, but not both.
> +  PropagationPolicy *DeletePropagationPolicy
> +}
> +
> +type DeletePropagationPolicy string
> +
> +const (
> +    // The default depends on the existing finalizers on the object and the type of the object.
>
> In the comment, note explicitly that the default depends on the object
> type and the controller for the object. (If the controller pre-adds the
> sync finalizer label, then you'll get foreground behavior as the default
> unless you specifically request something else).
>
> @lavalamp <https://github.com/lavalamp> Is the current comment clear? I
> think saying the default depends on the "existing finalizers" is more
> general than saying it depends on "the controller for the object".
> ------------------------------
>
> In docs/proposals/synchronous-garbage-collection.md
> <https://github.com/kubernetes/kubernetes/pull/32628#pullrequestreview-3981287>
> :
>
> > +
> +## DeleteOptions
> +
> +```go
> +DeleteOptions {
> +  …
> +  // Whether and how garbage collection will be performed.
> +  // Defaults to DeletePropagationDefault
> +  // Either this field or OrphanDependents may be set, but not both.
> +  PropagationPolicy *DeletePropagationPolicy
> +}
> +
> +type DeletePropagationPolicy string
> +
> +const (
> +    // The default depends on the existing finalizers on the object and the type of the object.
>
> Also, do we need a DeletePropagationDefault? Can we use nil to mean
> default?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32628#pullrequestreview-3981287>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngltsu4xlMUfmNf9xbTSbZzaf0eZShks5qzWvtgaJpZM4J8Saa>
> .
>

caesarxuchao commented:
@lavalamp can we merge the proposal? @deads2k had given his lgtm (https://github.com/kubernetes/kubernetes/pull/32628#issuecomment-253276985)

caesarxuchao commented:
Is there any objection merging the proposal? If not I'll apply the lgtm tomorrow.

caesarxuchao commented:
Applying the lgtm label.

There is not much time left for 1.5. I'll focus on converting controllers to use client-go. Probably I'll do the implementation of this in 1.6.
